README.rst
Python Client for Essential Contacts API API
============================================

|stable| |pypi| |versions|

`Essential Contacts API API`_: 

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-essential-contacts.svg
   :target: https://pypi.org/project/google-cloud-essential-contacts/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-essential-contacts.svg
   :target: https://pypi.org/project/google-cloud-essential-contacts/
.. _Essential Contacts API API: https://cloud.google.com/resource-manager/docs/managing-notification-contacts/
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/essentialcontacts/latest
.. _Product Documentation:  https://cloud.google.com/resource-manager/docs/managing-notification-contacts/

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Essential Contacts API API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Essential Contacts API API.:  https://cloud.google.com/resource-manager/docs/managing-notification-contacts/
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-essential-contacts


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-essential-contacts

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Essential Contacts API API
   to see other available methods on the client.
-  Read the `Essential Contacts API API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Essential Contacts API API Product documentation:  https://cloud.google.com/resource-manager/docs/managing-notification-contacts/
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

## [0.4.2](https://github.com/googleapis/python-essential-contacts/compare/v0.4.1...v0.4.2) (2022-06-07)


### Bug Fixes

* **deps:** require protobuf<4.0.0 on v0 branch ([#121](https://github.com/googleapis/python-essential-contacts/issues/121)) ([c6af494](https://github.com/googleapis/python-essential-contacts/commit/c6af4949576bf273809d61174e98bf83e65c845c))

### [0.4.1](https://github.com/googleapis/python-essential-contacts/compare/v0.4.0...v0.4.1) (2022-04-01)


### Bug Fixes

* **deps:** require google-api-core >= 1.31.5, >= 2.3.2 on v0 release ([#96](https://github.com/googleapis/python-essential-contacts/issues/96)) ([1973650](https://github.com/googleapis/python-essential-contacts/commit/1973650bc27bca1b6fdb96f566454870631353d5))

## [0.4.0](https://www.github.com/googleapis/python-essential-contacts/compare/v0.3.0...v0.4.0) (2021-10-18)


### Features

* add support for python 3.10 ([#54](https://www.github.com/googleapis/python-essential-contacts/issues/54)) ([74cb082](https://www.github.com/googleapis/python-essential-contacts/commit/74cb0823f8fd16b198c4de71c5df4105736e2d18))

## [0.3.0](https://www.github.com/googleapis/python-essential-contacts/compare/v0.2.4...v0.3.0) (2021-10-08)


### Features

* add context manager support in client ([#51](https://www.github.com/googleapis/python-essential-contacts/issues/51)) ([87911bc](https://www.github.com/googleapis/python-essential-contacts/commit/87911bcb578511d49b1263fc4e3c1c74c328d59a))

### [0.2.4](https://www.github.com/googleapis/python-essential-contacts/compare/v0.2.3...v0.2.4) (2021-10-04)


### Bug Fixes

* improper types in pagers generation ([831f18a](https://www.github.com/googleapis/python-essential-contacts/commit/831f18afd43d83310699352e7467641c4dc9a79a))

### [0.2.3](https://www.github.com/googleapis/python-essential-contacts/compare/v0.2.2...v0.2.3) (2021-09-24)


### Bug Fixes

* add 'dict' annotation type to 'request' ([9fdb6bf](https://www.github.com/googleapis/python-essential-contacts/commit/9fdb6bfa9457ec4faf0463920c08c5eb2bd1407b))

### [0.2.2](https://www.github.com/googleapis/python-essential-contacts/compare/v0.2.1...v0.2.2) (2021-07-29)


### Bug Fixes

* enable self signed jwt for grpc ([#27](https://www.github.com/googleapis/python-essential-contacts/issues/27)) ([fdbaa0e](https://www.github.com/googleapis/python-essential-contacts/commit/fdbaa0e78a35657c36b6ec25dbea113cd18e98b0))


### Documentation

* add Samples section to CONTRIBUTING.rst ([#23](https://www.github.com/googleapis/python-essential-contacts/issues/23)) ([a3e6040](https://www.github.com/googleapis/python-essential-contacts/commit/a3e6040d854e213620031c65276997dd39bce428))


### Miscellaneous Chores

* release as 0.2.2 ([#28](https://www.github.com/googleapis/python-essential-contacts/issues/28)) ([e1ec288](https://www.github.com/googleapis/python-essential-contacts/commit/e1ec2887af5a87395c7af79e07189b35ee13d85f))

### [0.2.1](https://www.github.com/googleapis/python-essential-contacts/compare/v0.2.0...v0.2.1) (2021-07-21)


### Bug Fixes

* **deps:** pin 'google-{api,cloud}-core', 'google-auth' to allow 2.x versions ([#22](https://www.github.com/googleapis/python-essential-contacts/issues/22)) ([926e1ea](https://www.github.com/googleapis/python-essential-contacts/commit/926e1ea935ada48c24cef48df201e279e692e97a))

## [0.2.0](https://www.github.com/googleapis/python-essential-contacts/compare/v0.1.1...v0.2.0) (2021-07-14)


### Features

* add always_use_jwt_access ([#14](https://www.github.com/googleapis/python-essential-contacts/issues/14)) ([66e7ee0](https://www.github.com/googleapis/python-essential-contacts/commit/66e7ee0f82527e33c708949088493ad90e634fb1))


### Bug Fixes

* disable always_use_jwt_access ([#18](https://www.github.com/googleapis/python-essential-contacts/issues/18)) ([fcc8156](https://www.github.com/googleapis/python-essential-contacts/commit/fcc815666514d9827c12507a4e601a409bce2235))


### Documentation

* omit mention of Python 2.7 in 'CONTRIBUTING.rst' ([#1127](https://www.github.com/googleapis/python-essential-contacts/issues/1127)) ([#9](https://www.github.com/googleapis/python-essential-contacts/issues/9)) ([31ae5a1](https://www.github.com/googleapis/python-essential-contacts/commit/31ae5a198e8844bfcc57283da8b26bcef17d6814)), closes [#1126](https://www.github.com/googleapis/python-essential-contacts/issues/1126)

### [0.1.1](https://www.github.com/googleapis/python-essential-contacts/compare/v0.1.0...v0.1.1) (2021-06-16)


### Bug Fixes

* exclude docs and tests from package ([#6](https://www.github.com/googleapis/python-essential-contacts/issues/6)) ([40c2402](https://www.github.com/googleapis/python-essential-contacts/commit/40c2402563118085e92cf9039e5f7be953ae57e8))

## 0.1.0 (2021-06-02)


### Features

* generate v1 ([4a7876b](https://www.github.com/googleapis/python-essential-contacts/commit/4a7876b5636917bae00d5947bf837bdfe042adfa))


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.6",
    "3.7",
    "3.8",
    "3.9",
    "3.10",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==19.10b0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.9"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'asyncmock', 'pytest-asyncio')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/essential_contacts_v1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
